sympy/core/basic.py

```diff
@@ -367,49 +367,106 @@ def __ne__(self, other):
 
     def dummy_eq(self, other, symbol=None):
         """
-        Compare two expressions and handle dummy symbols.
+        Return True if ``self`` and ``other`` are the same (or
+        can be made so by mapping Dummy symbols in ``self`` to
+        Dummy symbols (default) in ``other``. Symbols that are
+        not actually Dummy symbols in ``other`` can be treated
+        as such by passing them -- one or more -- in ``symbol``.
+
+        If there are no Dummy symbols in ``self`` the result
+        can only be true if the expressions are structurally the
+        same in their use of free symbols.
 
         Examples
         ========
 
-        >>> from sympy import Dummy
-        >>> from sympy.abc import x, y
+        >>> from sympy import Dummy, Sum, symbols
+        >>> from sympy.abc import x, y, z
+        >>> _d, _e, _f = symbols('d e f', cls=Dummy)
 
-        >>> u = Dummy('u')
+        Normal equality tests fail if the symbols are not
+        the same:
 
-        >>> (u**2 + 1).dummy_eq(x**2 + 1)
-        True
-        >>> (u**2 + 1) == (x**2 + 1)
+        >>> _d == _e
         False
 
-        >>> (u**2 + y).dummy_eq(x**2 + y, x)
+        This even applies for most objects that use bound symbols:
+
+        >>> Sum(x, (x, 1, 2)) == Sum(y, (y, 1, 2))
+        False
+
+        Both expressions will test the same with this function which
+        returns True if Dummy symbols in ``self`` can be paired
+        uniquely to Dummy symbols in ``other``:
+
+        >>> (_d + 1).dummy_eq(_e + 1)
         True
-        >>> (u**2 + y).dummy_eq(x**2 + y, y)
+
+        Non-Dummy symbols in ``other`` are treated like constants
+        and will not match to a Dummy in ``self`` by default:
+
+        >>> (_d + 1/x).dummy_eq(y + 1/x)
         False
 
-        """
-        s = self.as_dummy()
-        o = _sympify(other)
-        o = o.as_dummy()
+        But to treat a given symbol like a Dummy, pass it as
+        ``symbol``:
 
-        dummy_symbols = [i for i in s.free_symbols if i.is_Dummy]
+        >>> (_d + 1/x).dummy_eq(y + 1/x, y)
+        True
+        >>> (_d + _e*(1 + x)).dummy_eq(y + z*(1 + x), (y, z))
+        True
 
-        if len(dummy_symbols) == 1:
-            dummy = dummy_symbols.pop()
-        else:
-            return s == o
+        Notes
+        =====
 
-        if symbol is None:
-            symbols = o.free_symbols
+        It is possible to see if any two expressions are structurally
+        the same with this function by passing True for ``symbol``:
 
-            if len(symbols) == 1:
-                symbol = symbols.pop()
-            else:
-                return s == o
+        >>> a = x + 2*y
+        >>> b = y + 2*x
+        >>> a.dummy_eq(b, True)
+        True
```

Comment:
To me this seems like a separate function from dummy_eq that should have a different name.

Reply:
Would you be in favor of putting `dummy_eq` on a deprecation path and replacing it with `sym_eq` which will tell if `self` and `other` are symbolically the same in specified symbols, defaulting to all?

```python
x.sym_eq(y)
Sum(x, (x, 1, 2)).sym_eq(Sum(y, (y, 1, 2)))
(x + y).sym_eq(x + z, z)
(Dummy() + y).sym_eq(x + y)
(Dummy() + y).sym_eq(x + y, x) # y->y, x->Dummy()
not (Dummy() + y).sym_eq(x + y, y)  # nothing for x to match to
```

Reply:
I'm not sure I understand what you propose.

What is the reason for wanting to change `dummy_eq`?

Reply:
OK, I made `dummy_eq` to return true for only 3 cases:

1) _aresame(self, other)
2) Dummy in self maps to Dummy in other (default)
3) Dummy in self maps to indicated symbols in other

Reply:
This breaks the hopefully-not-used `x.dummy_eq(Dummy())` syntax.